data_source_examples/aha/using_pyairtable/example.py

```diff
@@ -0,0 +1,127 @@
+
+features = getFeatures()
+print('All feature retrieved.')
+
+# Map feature details to Airtable field names
+def mapFeatures(feature):
+    return {
+        'Name': feature.get('name'),
+        'Aha ID': feature.get('id'),
+        'Reference Num': feature.get('reference_num'),
+        'Created At': feature.get('created_at'),
+        'URL': feature.get('url'),
+        'Product ID': feature.get('product_id')
+    }
+
+inputRecords = list(map(mapFeatures, features))
+
+# For each feature, get additional details through another API call
+print('\nGetting additional feature details. This may take several moments.')
+for record in inputRecords:
+    featureEndpoint = endpoint + '/' + record.get('Reference Num')
+    featureResponse = requests.get(featureEndpoint, headers=headers).json()
+    feature = featureResponse.get('feature')
+    record['Start Date'] = feature.get('start_date')
+    record['End Date'] = feature.get('due_date')
+    record['Progress'] = feature.get('progress') * .01 if feature.get('progress') else None
+    record['Progress Source'] = feature.get('progress_source')
+print('Feature details retrieved')
+
+# Retrieve all existing records from the base through the Airtable REST API
+allExistingRecords = Table.all()
+
+# Create an object mapping of the primary field to the record ID
+# Remember, it's assumed that the AIRTABLE_UNIQUE_FIELD_NAME field is truly unique
+upsertFieldValueToExistingRecordId = {
+    existingRecord['fields'].get(AIRTABLE_UNIQUE_FIELD_NAME): existingRecord['id'] for existingRecord in allExistingRecords
+}
+
+# Create two arrays: one for records to be created, one for records to be updated
+recordsToCreate = []
+recordsToUpdate = []
+
+# For each input record, check if it exists in the existing records. If it does, update it. If it does not, create it.
+for inputRecord in inputRecords:
+    recordUniqueValue = inputRecord.get(AIRTABLE_UNIQUE_FIELD_NAME, None)
+    # print('\tProcessing record w / \'{}\' === \'{}\''.format(
+    #     AIRTABLE_UNIQUE_FIELD_NAME, recordUniqueValue))
+
+    existingRecordIdBasedOnUpsertFieldValueMaybe = upsertFieldValueToExistingRecordId.get(
+        recordUniqueValue)
+
+    # and if the upsert field value matches an existing one...
+    if existingRecordIdBasedOnUpsertFieldValueMaybe:
+        # Add record to list of records to update
+        recordsToUpdate.append(
+            dict(id=existingRecordIdBasedOnUpsertFieldValueMaybe, fields=inputRecord))
+    else:
+        # Otherwise, add record to list of records to create
+        recordsToCreate.append(inputRecord)
+
+# Read out array sizes
+print("\n{} records to create".format(len(recordsToCreate)))
+print("{} records to update".format(len(recordsToUpdate)))
+
+# Perform record creation
+Table.batch_create(recordsToCreate)
+
+# Perform record updates on existing records
+Table.batch_update(recordsToUpdate)
+
+# Uncomment the following block of code if you wish to set records in Airtable # that do not exist in
```

Comment:
Is this specific to Aha? wasn't really following what the purpose of this is

Reply:
This is something that Mark added a little while to include in upsert examples. The idea is that if the user wants to identify records that exist in Airtable, but no longer exist in the other data source, this allows them to identify those easily.